src/processor.cpp

```diff
@@ -0,0 +1,413 @@
+    int fuel_intake;
+    int duration;
+    //TODO: Skill used
+
+    bool was_loaded;
+    void load( JsonObject &jo, const std::string &src );
+    void check() const;
+    void reset();
+
+};
+
+process_id get_process_id( std::string id )
+{
+    return process_id( id );
+}
+
+generic_factory<process_data> processes_data( "process type", "id" );
+
+}
+
+void process_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "name", name );
+    //TODO: modify to allow lists of components
+    mandatory( jo, was_loaded, "components", components,  string_id_reader<itype_id> {} );
+    mandatory( jo, was_loaded, "output", output, string_id_reader<itype_id> {} );
+    mandatory( jo, was_loaded, "fuel_intake", fuel_intake );
+    mandatory( jo, was_loaded, "duration", duration );
+}
+
+void process_data::check() const
+{
+    const process_id proc_id( id.str() );
+
+    if (!proc_id.is_valid()) {
+        debugmsg("There is no process with the name %s", id.c_str());
+    }
+    /*
+    I'm not sure if I need this, it would validate components and output as valid itype_id
+    for( itype_id elem : components ) {
+        //The compiler complains about this is_valid method
+        if( !elem.is_valid() ) {
+            debugmsg( "Invalid components \"%s\" in \"%s\".", elem.c_str(), id.c_str() );
+        }
+    }
+    if( !output.is_valid() ) {
+        debugmsg( "Invalid output \"%s\" in \"%s\".", output.c_str(), id.c_str() );
+    }
+    */
+    if( !fuel_intake >= 0 ) {
+        debugmsg( "Invalid fuel intake \"%d\" in \"%s\".", fuel_intake, id.c_str() );
+    }
+    if( !duration > 0 ) {
+        debugmsg( "Invalid duration \"%d\" in \"%s\".", duration, id.c_str() );
+    }
+}
+
+void process_data::load( JsonObject &jo, const std::string &src )
+{
+    processes_data.load( jo, src );
+}
+
+void process_data::check() const
+{
+    processes_data.check();
+}
+
+void process_data::reset()
+{
+    processes_data.reset();
+}
+
+template <typename T>
+bool IsSubset(std::vector<T> A, std::vector<T> B)
+{
+    std::sort(A.begin(), A.end());
+    std::sort(B.begin(), B.end());
+    return std::includes(A.begin(), A.end(), B.begin(), B.end());
+}
+
+process_data select_active(const tripoint &examp, player &p)
+{
+    processor_id pid = get_processor_id(examp);
+    const processor_data &current_processor = processors_data.obj(pid);
+    //Make an object list of processes
+    std::vector<process_data> possible_processes;
+    for (process_id &p : current_processor.processes) {
+        possible_processes.push_back(processes_data.obj(p));
+    }
+    //Figure out active process
+    process_data* active_process;
+    if (possible_processes.size() == 1) {
+        *active_process = possible_processes[0];
+    }
+    else {
+        map_stack items = g->m.i_at( examp );
+        if (!items.empty()) {
+            //Build a set of item_id's present
+            std::vector<itype_id> present_items;
+            for (item i : items) {
```

Comment:
You should use references here:
```
for( const item &i : items ) {
```
Otherwise you are creating copies of the items just to check the ids.